### ISSUE TITLE
Recursively binding Multipart form data to complex nested request DTOs

### ISSUE THREAD

**Issue by kevinkouketsu** [FastEndpoints/FastEndpoints#782]
I have this request:
```c#
public record Request
{
    public IList<MediaWithDescription> Medias { get; set; }
    
    public class MediaWithDescription
    {
        public IFormFile Media { get; set; }
        public string? Description { get; set; }
    }
}
```

Normally, I would add [FromForm] to get this correctly in an ASP.Net Core MVC project. It works fine. 
For example:
```c#
    [HttpPost("test-upload/")]
    [TranslateResultToActionResult]
    [AllowAnonymous]
    [Consumes("multipart/form-data")]
    public async Task<Result> Test([FromForm] Request request)
    {
        return Result.Success();
    }
```
However, when using FastEndpoints, it doesn't work—Medias is always null

```c#

public class UploadMediaPet : Endpoint<Request, Results<NoContent, ProblemHttpResult>>
{
    public override void Configure()
    {
        Post("/api/pet/{PetId}/medias/");
        AllowFileUploads();
    }

    public override Task<Results<NoContent, ProblemHttpResult>> ExecuteAsync([FromForm] Request req, CancellationToken ct)
    {
        return base.ExecuteAsync(req, ct);
    }
}

```
What should I do in this case? Should I handle it manually?

**Comment by dj-nitehawk** [comment]
binding form data to complex nested dtos is currently not supported.
as of now, only the root level can be automatically bound.

going to need the help of the community to implement this one as it's a bit of a complicated task and i don't have the capacity right now to look at it.

basically need to refactor [this method](https://github.com/FastEndpoints/FastEndpoints/blob/c1218340324e4afb3d4f61ff564fbaedda534e70/Src/Library/Binder/RequestBinder.cs#L231-L298) to be able to recursively bind deeply nested dtos.

---

for the time being, you'll have to resort to a regular minimal api endpoint as they seem to support this scenario from .net 8 onwards.

**Comment by kevinkouketsu** [comment]
I'll try to implement this when I get the chance.

Thanks for the clarification.

**Comment by dj-nitehawk** [comment]
for anyone looking in to this, the multipart form binding code should be able to correctly handle a complex request dto object graph such as the following:

```cs
class Book
{
    public string Title { get; set; }
    public IFormFile CoverImage { get; set; }
    public IFormFileCollection SourceFiles { get; set; }
    public Author MainAuthor { get; set; }
    public List<Author> CoAuthors { get; set; }
    public Address PublisherAddress { get; set; }
    public List<Address> RetailerAddresses { get; set; }
}

class Author
{
    public string Name { get; set; }
    public IFormFile ProfileImage { get; set; }
    public IFormFileCollection DocumentFiles { get; set; }
    public Address MainAddress { get; set; }
    public List<Address> OtherAddresses { get; set; }
}

class Address
{
    public string Street { get; set; }
    public IFormFile MainImage { get; set; }
    public IFormFileCollection AlternativeImages { get; set; }
}
```

**Comment by dj-nitehawk** [comment]
added an integration test to work against:

https://github.com/FastEndpoints/FastEndpoints/blob/49e85726c0ff6730bfc6e0b0f2ebbb22c20e5fc7/Tests/IntegrationTests/FastEndpoints/BindingTests/BindingTests.cs#L674-L726

**Comment by dj-nitehawk** [comment]
can you give `v5.30.0.4-beta` a try?

i've done a [crude implementation](https://github.com/FastEndpoints/FastEndpoints/blob/main/Src/Library/Binder/ComplexFormBinder.cs).
there may be edge cases that i haven't thought of and there's definitely room for improvement.
since my capacity is limited, i'd like the help of the community to optimize that class.

---

have a look at the [integration test](https://github.com/FastEndpoints/FastEndpoints/blob/4a589ad9bfa8601122aff1c773cab35b75d4efdc/Tests/IntegrationTests/FastEndpoints/BindingTests/BindingTests.cs#L662) and the [dto structure](https://github.com/FastEndpoints/FastEndpoints/blob/4a589ad9bfa8601122aff1c773cab35b75d4efdc/Web/%5BFeatures%5D/TestCases/Binding/FormBindingComplexDtos/Endpoint.cs#L3-L7) to get an idea of how to use.

basically, just like the `[FastEndpoints.FromBody]` attribute usage, use the `[FastEndpoints.FromForm]` attribute to trigger binding complex data from forms.

**Comment by kevinkouketsu** [comment]
I am testing the `v5.30.0.4-beta` and still having issues.

```
public record Request
{
    [FromForm]
    public List<MediaWithDescription> Medias { get; set; }
    
    public class MediaWithDescription
    {
        public IFormFile Media { get; set; }
        public string? Description { get; set; }
    }
}

Pretty much the DTO you provided. 

![image](https://github.com/user-attachments/assets/ca1a9260-ae1a-4ffa-b627-f307ce280c47)

```
System.Reflection.TargetParameterCountException: Parameter count mismatch.
   at System.Reflection.RuntimeMethodInfo.InvokePropertySetter(Object obj, BindingFlags invokeAttr, Binder binder, Object parameter, CultureInfo culture)
   at System.Reflection.RuntimePropertyInfo.SetValue(Object obj, Object value, Object[] index)
   at FastEndpoints.ComplexFormBinder.BindPropertiesRecursively(Object obj, IFormCollection form, String prefix)
   at FastEndpoints.ComplexFormBinder.Bind(PropCache fromFormProp, Object requestDto, IFormCollection forms)
   at FastEndpoints.RequestBinder`1.<BindFormValues>g__Execute|27_0(<>c__DisplayClass27_0&)
   at FastEndpoints.RequestBinder`1.BindFormValues(TRequest req, HttpRequest httpRequest, List`1 failures, Boolean dontAutoBindForm)
   at FastEndpoints.RequestBinder`1.BindAsync(BinderContext ctx, CancellationToken cancellation)
```

**Comment by dj-nitehawk** [comment]
it has to be wrapped in a class like this:
```cs
public class Request
{
    [FromForm]
    public FormBody Form { get; set; }

    public class FormBody
    {
        public List<MediaWithDescription> Medias { get; set; }

        public class MediaWithDescription
        {
            public IFormFile Media { get; set; }
            public string? Description { get; set; }
        }
    }
}
```

`v5.30.0.5-beta` will correctly warn you if used incorrectly.

**Comment by kevinkouketsu** [comment]
It worked. Awesome

**Comment by dj-nitehawk** [comment]
next version of FE ships with a source generator for reflection based methods which should help with this.
